### ISSUE TITLE
Give bt Adapter a name

### ISSUE THREAD

**Issue by dingausmwald** [custom-components/ble_monitor#830]
### What happened?

It happend again, after a reboot/power off, my bt adapters switched order (hci0 to hci1).
Since they are only recognized by hci order and named by mac address and i dont remember them after some months, i tried everything to bring back some devices as they seemed out of range, changed batteries and stuff. After i tried everything, i did the last, activated the other hci and well, there we are.
It seems like the setting itself is fixed to the hci0 or 1, so changing order after reboot is a problem. Could this be to mac, or even better, a name that can be set by the user? Like, "internal bt" and "bluetooth stick"? The internal bt adapter gets badly jamed by usb3 (ssd setup).

Or do i have to disable the internal bt adapter at the end?

**Comment by Ernst79** [comment]
How did you configure ble monitor? 

In YAML you can configure it based on MAC address like this.

```yaml
ble_monitor:
  bt_interface:
    - '04:B1:38:2C:84:2B'
```

in UI configuration, you can also set it based on its Mac address

**Comment by dingausmwald** [comment]
I did by UI. I am quite sure I didn't change anything, and suddenly it took the other adapter. What I could read hci0 or 1 gets set depending on the boot order of the adapters. Sometimes it can switch. Are you sure it's set by mac and not hci#?

**Comment by Ernst79** [comment]
Yes, I’m pretty sure, at least it should. However, you have to open the UI configuration one time to save the MAC address. Till that, it will use the first hci.

**Comment by dingausmwald** [comment]
Hey, i had some time to invest again because the problem is still there. Summerized by claude

Following up on this — I don't think I ever said the *initial* MAC resolution was broken. My original point was specifically about what happens when the adapter crashes/gets reset mid-session and the kernel renumbers the hci indices — not about the config-flow resolving to the wrong adapter at setup time.

I just reproduced this cleanly and can point to the exact code path this time.

**Setup**
- ble_monitor 13.14.0, configured via UI (`is_flow: true`), `bt_interface: ["00:E0:43:76:3E:03"]`
- Two adapters: onboard Pi BT (`E4:5F:01:74:E1:FF`) and a Realtek RTL8761B USB dongle (`0bda:8771`, MAC `00:E0:43:76:3E:03`) — the latter is the one ble_monitor is configured to use.

**What happened**
The Realtek dongle hit a hardware fault and got reset by the kernel:

```
Bluetooth: hci2: Received unexpected HCI Event 0x00
Bluetooth: hci2: unexpected event 0x01 length: 0 < 1
Bluetooth: hci2: RTL: hw err, trigger devcoredump (24)
...
Bluetooth: hci2: command 0x0c03 tx timeout
Bluetooth: hci2: HCI reset during shutdown failed
Bluetooth: hci2: Resetting usb device.
usb 1-1.4.1: reset full-speed USB device number 4 using xhci_hcd
Bluetooth: hci0: RTL: examining hci_ver=0a hci_rev=dfc6 lmp_ver=0a lmp_subver=d922
...
```

After the reset, the system only has 2 controllers left (hci0/hci1) — the dongle that used to be `hci2` came back under a different index. From then on, every scan cycle logs:

```
HCIdump thread: OS error (hci2): [Errno 19] error while attempting to bind on interface 2: No such device
HCIdump thread: Trying to power cycle Bluetooth adapter hci2 00:E0:43:76:3E:03, will try to use it next scan period.
bt_helpers: RF kill switch check failed - no data for hci2. Available data: {...}
bt_helpers: hci2 seems not to exist (anymore), check BT interface mac address in your settings. Available adapters: {0: '00:E0:43:76:3E:03', 1: 'E4:5F:01:74:E1:FF'}
```

...repeating every ~60s indefinitely. No BLE data comes in again until Home Assistant Core is fully restarted (a plain integration reload doesn't help either).

**Root cause (traced through the current source)**

1. `BT_INTERFACES` in `bt_helpers.py` is a module-level constant computed once — `BT_INTERFACES = hci_get_mac([0, 1, 2, 3])` — at Python import time. It's never refreshed for the lifetime of the HA process.
2. The MAC→hci resolution in `__init__.py` (the part that actually honors `bt_interface` by MAC) only runs once, in `async_setup_entry`. That's the part that "works", and is what the earlier answers in this thread and in #1044 were about.
3. The runtime recovery path — `reset_bluetooth()` in `bt_helpers.py`, invoked from the "Trying to power cycle..." log line — is called with the *stale numeric hci index* captured at setup time. It calls `MGMTBluetoothCtl(hci)` fresh each time (so it *is* live-querying the kernel), but by index, not by the MAC it already has stored in `bt_mac_list`. When that index no longer exists, `adapter.mac` comes back `None`, it logs the "seems not to exist" error, and gives up — it never re-looks-up the current index for the known MAC.

So: MAC-based selection at config time is correct and does work as advertised. The gap is that the failure-recovery path doesn't re-run that same MAC lookup after a live adapter crash/renumber — it just keeps retrying a dead index forever.

**Re-opening/saving the UI config doesn't help either**

Since the earlier advice in this thread was "you have to open the UI configuration one time to save the MAC address", I tried exactly that today, after the failure was already ongoing. It didn't change anything, and I think I can show why:

```
modified_at: 2026-07-08T17:49:01
hci_interface: [2]
bt_interface: ['00:E0:43:76:3E:03']
```

That's the config entry's stored `options`, read straight from `.storage/core.config_entries` right after I saved via the UI today (2026-07-15). The `modified_at` timestamp is a week old — my save didn't rewrite it.

The reason is the same root cause as above, one level up: `__init__.py` does `from .bt_helpers import BT_INTERFACES` at import time. That's a one-time name binding — it doesn't re-run `hci_get_mac()` on every options-flow save or integration reload, it just reuses whatever `bt_helpers.BT_INTERFACES` was when this HA process last imported the module (i.e. the last full Core restart). The adapter picker shown in the UI does query live and shows the correct current adapters, which is presumably why the advice sounded right — but that's a separate code path from the one that actually resolves and stores the hci index used by the scanner. Saving in the UI re-runs the resolution logic against the same frozen lookup table, so at best it reproduces the same (by-now wrong) answer; it can't discover that the adapter moved.

Concretely: `hci_interface: [2]` in the stored options was *correct* on 2026-07-08 — that's genuinely when the dongle sat at hci2, confirmed by dmesg. It just never got refreshed after the adapter later crashed again and came back at a different index, and no amount of re-opening/saving the config flow reruns that detection — only a full Home Assistant Core restart does (which re-imports the module and calls `hci_get_mac()` fresh).

**Suggested fix**

In the `OSError`/`reset_bluetooth` path, when the configured hci index is reported missing, re-run `hci_get_mac([0,1,2,3])` and look up the current index for the already-known MAC (`bt_mac_list`) instead of only retrying the frozen index. That would make it self-heal after a dongle crash without requiring a full HA Core restart.

Happy to test a beta/patch against this exact setup (dongle fault is reproducible on my Pi every few days) if that's useful.
